envier/env.py

```diff
@@ -243,20 +294,44 @@ def __init__(self, source=None, parent=None):
         derived = []
         for name, e in list(self.__class__.__dict__.items()):
             if isinstance(e, EnvVariable):
-                setattr(self, name, e(self, self._full_prefix))
+                s, v = e(self, self._full_prefix)
+                self.set_attr_source_value(name, s, v)
             elif isinstance(e, type) and issubclass(e, Env):
                 if e.__item__ is not None and e.__item__ != name:
                     # Move the subclass to the __item__ attribute
                     setattr(self.spec, e.__item__, e)
                     delattr(self.spec, name)
                     name = e.__item__
-                setattr(self, name, e(source, self))
+                setattr(self, name, e(source=self.source, parent=self))
             elif isinstance(e, DerivedVariable):
                 derived.append((name, e))
 
         for n, d in derived:
             setattr(self, n, d(self))
 
+    def __setattr__(self, name, value):
+        if name != "_items":
+            self._items[name].set_source("programmatic", value)
+        super(Env, self).__setattr__(name, value)
```

Comment:
I love the general idea of keeping track of how a value was set. I just wonder if we could make use of descriptors and dynamic classes here to tidy the logic a bit. For example:

```python
from enum import Enum


class Source(Enum):
    DEFAULT = 0
    OVERRIDE = 1


class VarInstance:
    def __init__(self, *args, **kwargs):
        try:
            super().__init__(*args, **kwargs)
        except Exception:
            pass
        self.source = Source.DEFAULT

    # Add support for internal mutation
    def append(self, *args, **kwargs):
        self.source = Source.OVERRIDE
        return super().append(*args, **kwargs)


def make_var_instance(_type):
    return type(f"VarInstance[{_type.__name__}]", (VarInstance, _type), {})


# These should be built on demand
IntVarInstance = make_var_instance(int)
ListVarInstance = make_var_instance(list)

# Descriptor wrapper to keep track of the source
class Var:
    def __init__(self, value):
        self._value = value
        self._source = None

    def __get__(self, obj, objtype=None):
        return self._value

    def __set__(self, obj, value):
        self._value = IntVarInstance(value)
        self._value.source = Source.OVERRIDE


class Config:
    var = Var(IntVarInstance(10))
    lst = Var(ListVarInstance([1, 2, 3]))


config = Config()
assert config.var == 10, config.var
assert config.var.source == Source.DEFAULT
assert config.lst == [1, 2, 3]
assert config.lst.source == Source.DEFAULT

config.var = 20
config.lst.append(4)
assert config.var == 20
assert config.var.source == Source.OVERRIDE
assert config.lst == [1, 2, 3, 4]
assert config.lst.source == Source.OVERRIDE
```

IIUC, one thing that this change doesn't handle is the case of internal mutation. That is, if a list is mutated programmatically, we don't set the source accordingly (should we?).

Reply:
@P403n1x87 is this something you would like to do right now or can we leave it for a future refactor? I prefer the simplicity at the moment since we're still fleshing out the feature

Reply:
Ah it would be hard for me to give an "objective" answer here, as TBH I think the descriptor + dynamic types approach is simpler and leaner in my head (and that's why I proposed it). So my honest answer is that I'd prefer we went with this other approach, if we can make it work. Perhaps you might want to get some more eyes on this PR from e.g. the guild so that you could have more opinions besides my very polarised one 🙂 .